src/transformers/astro.ts

```diff
@@ -0,0 +1,44 @@
+import { UrlParser, UrlTransformer } from "../types.ts";
+import {
+  getNumericParam,
+  setParamIfDefined,
+  toUrl,
+} from "../utils.ts";
+
+export interface AstroParams {
+  href: string;
+  quality?: string | number;
+}
+
+export const parse: UrlParser<AstroParams> = (url) => {
+  const parsedUrl = toUrl(url);
+  const href = encodeURIComponent(parsedUrl.searchParams.get("href") ?? new URL(parsedUrl.pathname, parsedUrl.origin).toString())
+  const width = getNumericParam(parsedUrl, "w") || undefined;
+  const height = getNumericParam(parsedUrl, "h") || undefined;
+  const format = parsedUrl.searchParams.get("f") || undefined;
+  const quality = parsedUrl.searchParams.get("q") || undefined;
+
+  return {
+    width,
+    height,
+    format,
+    base: `/_image?href=${href}`,
+    params: { quality, href },
+    cdn: "astro",
+  };
+};
+
+export const transform: UrlTransformer = (
+  { url: originalUrl, width, height, format, },
+) => {
+  const parsedUrl = toUrl(originalUrl);
+  const href = encodeURIComponent(new URL(parsedUrl.pathname, parsedUrl.origin).toString())
```

Comment:
You don't need to use `encodeURIComponent` because you're setting it in a URLSearchParams object. However you _do_ need to pass the URL to `toCanonicalUrl` from the utils file, in case the URL is relative.

Reply:
The results I get from the tests show that the url is not encoded anymore:
- original test result `{"base":"/_image?href=https%3A%2F%2Fimages.ctfassets.net%2Faaaa%2Fxxxx%2Fyyyy%2Fhow-to-wow-a-customer.jpg","params":{"href":"https%3A%2F%2Fimages.ctfassets.net%2Faaaa%2Fxxxx%2Fyyyy%2Fhow-to-wow-a-customer.jpg"},"cdn":"astro"}`
- new test result `{"base":"/_image?href=https://images.ctfassets.net/aaaa/xxxx/yyyy/how-to-wow-a-customer.jpg","params":{"href":"https://images.ctfassets.net/aaaa/xxxx/yyyy/how-to-wow-a-customer.jpg"},"cdn":"astro"}`
- example from Astro `/_image?href=https%3A%2F%2Fmedia.npr.org%2Fassets%2Fimg%2F2017%2F06%2F03%2Fgettyimages-453879976-59801bc8f1f6bda5175c3098bb0d9a28696394be.jpg&w=1915&f=webp`

I think the `encodeURIComponent` may still be necessary

Reply:
Also, I feel like `new URL(decodedHref)` would throw an error if the `decodedHref` is not an absolute url.  I’m not sure how to get around that while still using `toCanonicalUrlString`

Reply:
Other than this comment, I’ve resolved everything else for merge 😄 Please let me know what you think the best way to solve this issue is

Reply:
I think I solved it with the `toUrl` util and found the double encoding errors 🚀 should be good to go now